### PR TITLE
[Inductor] Relax size constraints for re-inplacing

### PR DESCRIPTION
**Pull request by BoyuanFeng**
Current reinplacing requires input buffer and output buffer has exactly the same storage size. However, matmul padding may increase the tensor size slightly for better performance, which prevents reinplacing.

This PR changes the size constraints to be:
- input and output buffer have exactly the same symbolic expression for storage size (i.e., sympy str).
- it's statically known that 0.99 * input_size <= output_size <= input_size

### Apply on llm.c
See the reuse of `buf1`.
Before relaxing size requirements on re-inplacing: ([P1703512078](https://www.internalfb.com/phabricator/paste/view/P1703512078))

After relaxing size requirements on re-inplacing: ([P1703513053](https://www.internalfb.com/phabricator/paste/view/P1703513053))

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

**Comment by vadimkantorov** [comment]
I wonder, if there are there similar problems for convs? (which also often do padding)

File: test/inductor/test_torchinductor.py [hunk 1]
```diff
@@ -13112,6 +13112,192 @@ def wrapper(inp, weight):
             _, code = run_and_get_code(wrapper, inp, weight)
             self.assertTrue("in_out_ptr" in code[1])
 
+        @torch._functorch.config.patch("donated_buffer", True)
+        def test_donated_buffer_inplace_gpt(self):
+            # model implementation from llm.c:
+            # https://github.com/karpathy/llm.c/blob/master/train_gpt2.py
+            class NewGELU(nn.Module):
```
Comment (eellison) [inline test/inductor/test_torchinductor.py:13119]: Would be great if we could have a more minimal test here. Such as by being able to specify donated buffer for forward (which is generically useful). Any number of possible changes could make this no longer trigger. Potentially either by monkeypatching `get_donated_idxs`, or, making a public api - see this pr : https://github.com/pytorch/pytorch/commit/e0d6072f692ac55508cd4187548b389afc794856. 


At the very least - you need to force pad mm for this test to fire.

File: torch/_inductor/codegen/wrapper.py [hunk 2]
```diff
@@ -90,6 +90,34 @@ def buffer_reuse_key(node: BufferLike) -> ReuseKey:
...
+
+    if (
+        # NB: this is symbolic so that we don't try to reuse a buffer
+        # for s0 for s1, just because they happen to share the same
+        # size hint
+        sympy_str(input_size)
+        == sympy_str(output_size)
+    ) or (
+        # statically known that 0.99 * input_size <= output_size <= input_size
+        V.graph.sizevars.statically_known_geq(output_size, 0.99 * input_size)
```
Comment (eellison) [inline torch/_inductor/codegen/wrapper.py:113]: hmm, i would be a little more lenient, given that inplacing is a form of memory savings.. maybe .95 ? idk.

File: test/inductor/test_torchinductor.py [hunk 3]
```diff
@@ -13112,6 +13112,192 @@ def wrapper(inp, weight):
...
+                with ctx:
+                    loss = model(x, y)
+                loss.backward()
+
+            _, code = run_and_get_code(wrapper, x, y)
+
+            # checks re-inplacing happens if pad_mm happens. Extra if-branch is added
+            # since pad_mm is profiling-based and may not happen on certain machines
+            # for this specific test case.
+            if (
```
Comment (eellison) [inline test/inductor/test_torchinductor.py:13293]: can just add force_shape_padding..

**Review by eellison — APPROVED** [review]
looks good, just a couple small comments ! would be nice to see if we could make the unit test more of a unit test in follow up. shouldn't need the whole model.
